runtime/gc_vlhgc/SchedulingDelegate.cpp

```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+			overhead = midpointPct;
+		} else {
+			/* 
+			 * If pgc time is above the max pgc time, map high PGC time values as very very high overhead, in efforts to bring the PGC time down to tarokTargetMaxPauseTime
+			 * If pgc time is only slightly above tarokTargetMaxPauseTime, then there is only a very small overhead penalty, 
+			 * wheras being 2x higher than the target pause time leads to a significantly bigger penalty  
+			 */
+			double overheadCurve = pow(1.03, ((double)partialGcTimeMs - xmaxt)) + midpointPct - 1;
+			overhead = OMR_MIN(100.0, overheadCurve);
+		}
+		
+	} else {
+		/* 
+		 * Eden sizing logic is trying to keep hybrid overhead between xminpct and xmaxpct, while trying to respect xmaxt. 
+		 * In this situation, when pgc times are very high (above xmaxt), the overhead score needs to return a low number, suggesting contraction.
+		 * If partialGcTimeMs is less than half of xmaxt, eden can expand without any fear of getting close to xmaxt - the mapped cpu overhead here is > xmaxpct (suggesting eden expansion)
+		 * Ex: 20ms -> 12% (suggest expansion), 2000ms -> 0.00% (suggest contraction)
+		 */
+		double slope = (xmaxpct - xminpct) / ((xmaxt/2) - xmaxt);
+		overhead = (slope * (double)partialGcTimeMs) + ((2.0 * xmaxpct) - xminpct);
+		overhead = OMR_MAX(overhead, 0.0);	
+		/*  Expanding simply because pgc time is very small is not a good idea, so return xmaxpct, so that if the pgc cpu overhead wants to expand, only then eden expands */
+		overhead = OMR_MIN(overhead, xmaxpct);
+	}
+
+	return overhead;
+}
+
+double 
+MM_SchedulingDelegate::calculateHybridEdenOverhead(MM_EnvironmentVLHGC *env, uintptr_t partialGcTimeMs, double overhead)
+{
+	/* 
+	 * When trying to size eden, there is a delicate balance between pgc overhead (here, overhead is cpu %, or % of time that pgc is active 
+	 * versus inactive -> ex: pgc = 100ms, over 1000ms, overhead = 10%). In certain applications, with certain allocation patterns/liveness, 
+	 * pgc average time may be negatively impacted by growing eden unbounded. 
+	 * This function blends the pgc average time (whether it be the actual pgc historic time, or a "predicted" pgc pause time, is left up to the caller) with overhead (% of time gc is active relative to mutator).
+	 * This strikes a much better balance between pgc pause times, and gc cpu overhead, than if just cpu overhead was used.
+	 * 
+	 * By mapping a pgc time to a corresponding overhead (% of time gc is active relative to mutator), eden sizing logic can make a decision as to whether 
+	 * it wants to contract/expand, based on how much it will change the overhead and pgc times.
+	 */
+	double actualPGCOverheadWeight = 0.5;
+	assert(overhead >= 0.0 && overhead <= 1.0);
+	double pgcTimeOverhead = mapPgcTimeToPgcOverhead(env, partialGcTimeMs);
+	double hybridEdenOverheadHundredBased = (actualPGCOverheadWeight * (overhead * 100)) + ((1- actualPGCOverheadWeight) * pgcTimeOverhead);
```

Comment:
you should still be able to use M_Math::weightedAverage (although we use it typically for historical weighted averaging)